k8s/src/main/scala/io/buoyant/k8s/IngressCache.scala

```diff
@@ -0,0 +1,128 @@
+  import IngressCache._
+
+  val api = namespace match {
+    case Some(ns) => v1beta1.Api(apiClient).withNamespace(ns).ingresses
+    case None => v1beta1.Api(apiClient).ingresses
+  }
+
+  private[this] object Closed extends Throwable
+  private[this] val state = Var.async[IngressState](Activity.Pending) { state =>
+    val closeRef = new AtomicReference[Closable](Closable.nop)
+    val pending = api.get(None, None, None, retryIndefinitely = true).respond {
+      case Throw(e) => state.update(Activity.Failed(e))
+      case Return(ingressList) =>
+        val initState: Seq[IngressSpec] = ingressList.items.flatMap(item => mkItem(item))
+        state.update(Activity.Ok(initState))
+        val (stream, close) = api.watch(None, None, None)
+        closeRef.set(close)
+        val _ = stream.foldLeft(initState) { (ingresses, watchEvent) =>
+          val newState: Seq[IngressSpec] = watchEvent match {
+            case v1beta1.IngressAdded(a) => mkItem(a).map(item => ingresses :+ item).getOrElse(ingresses)
+            case v1beta1.IngressModified(m) => mkItem(m).map(item => ingresses.filterNot(i => isNameEqual(i, item)) :+ item).getOrElse(ingresses)
+            case v1beta1.IngressDeleted(d) => mkItem(d).map(item => ingresses.filterNot(i => isNameEqual(i, item))).getOrElse(ingresses)
+            case v1beta1.IngressError(e) =>
+              log.error("k8s watch error: %s", e)
+              ingresses
+          }
+          state.update(Activity.Ok(newState))
+          newState
+        }
+    }
+    Closable.make { t =>
+      pending.raise(Closed)
+      Closable.ref(closeRef).close(t)
+    }
+  }
+
+  private[this] val ingresses: Activity[Seq[IngressSpec]] = Activity(state)
+  private[this] def isNameEqual(x: IngressSpec, y: IngressSpec): Boolean = x.name == y.name && x.namespace == y.namespace
+  private[this] def mkItem(ingress: v1beta1.Ingress): Option[IngressSpec] = {
+    val namespace = ingress.metadata.flatMap(meta => meta.namespace).getOrElse("default")
+    ingress.spec match {
+      case Some(spec) =>
+        val paths = for (
+          spec <- ingress.spec.toSeq;
+          rules <- spec.rules.toSeq;
+          rule <- rules;
+          http <- rule.http.toSeq;
+          path <- http.paths
+        ) yield {
+          IngressPath(rule.host, path.path, namespace, path.backend.serviceName, path.backend.servicePort)
+        }
+
+        val fallback = spec.backend.map(b => IngressPath(None, None, namespace, b.serviceName, b.servicePort))
+        Some(IngressSpec(ingress.metadata.flatMap(_.name), ingress.metadata.flatMap(_.namespace), fallback, paths))
+      case None => None
+    }
+  }
+
+  def matchPath(hostHeader: Option[String], requestPath: String): Future[Option[IngressPath]] =
+    ingresses.map { cache: Seq[IngressSpec] =>
+      getMatchingPath(hostHeader, requestPath, namespace, cache)
+    }.values.toFuture.flatMap(Future.const)
```

Comment:
each time this is called, an observation is opened on the Activity until it produces a non-pending value, which then populates the Future, and closes the observation.  The result of this is that for each call to `matchPath` the Activity will be started and then stopped.  

Since we want identification to be fast, I think we should just hold the Activity open forever.  One way to do this would be to change the `ingresses` val to a lazy val and also hold open an observation:

```
private[this] lazy val ingresses: Activity[Seq[IngressSpec]] = {
  val act = Activity(state)
  val _ = act.states.respond(_ => ()) // register a listener forever to keep the Activity open
  act
}
```